prov/util/src/util_mr_cache.c

```diff
@@ -566,3 +566,67 @@ int ofi_open_mr_cache(uint32_t version, void *attr, size_t attr_len,
 	*fid = cache_fid;
 	return 0;
 }
+
+static ssize_t ofi_copy_mr_iov(struct ofi_mr **mr, const struct iovec *iov,
+		size_t iov_count, uint64_t iov_offset, void *buf,
+		size_t size, int dir)
```

Comment:
This function looks like it belongs better in hmem.c than the mr_cache

Reply:
I went back and forth but put it in the MR cache because adding it the hmem would require including ofi_mr.h which requires ofi_hmem.h for the handle size so there's a circular dependency. Do you want me to try to move it?

Reply:
I think it makes more sense in hmem.c.  Looking at just the first commit, can you just include ofi_mr.h instead of ofi_hmem.h?  You just need struct ofi_mr, right?

Reply:
I tried that but I get a lot of these
```
In file included from ./include/ofi_mr.h:51,                                                                           [130/1171]
                 from ./include/ofi_util.h:58,                                                                                   
                 from src/fabric.c:47:                          
./include/ofi_hmem.h:380:62: warning: 'struct ofi_mr' declared inside parameter list will not be visible outside of this definiti
on or declaration               
 ssize_t ofi_copy_from_mr_iov(void *dest, size_t size, struct ofi_mr **mr,                                                       
                                                              ^~~~~~                                                             
./include/ofi_hmem.h:383:35: warning: 'struct ofi_mr' declared inside parameter list will not be visible outside of this definiti
on or declaration               
 ssize_t ofi_copy_to_mr_iov(struct ofi_mr **mr, const struct iovec *iov,                                                         
```

Reply:
Sorry, my brain wasn't working.  I meant move struct ofi_mr into ofi_hmem.h.  ofi_mr.h already includes ofi_hmem.h

Reply:
You'd rather move ofi_mr into ofi_hmem instead of have the mr copy function in the MR code? (That came out sarcastic but I'm just clarifying)

Reply:
Between those 2 options, yes.  But maybe we need to back up.  A better option would leave struct ofi_mr in ofi_mr.h, put the new calls in ofi_hmem.h, and have ofi_hmem.h include ofi_mr.h.  To do that we need to figure out why ofi_mr.h includes ofi_hmem.h.

At a quick glance, I think it's only to pick up the definition of MAX_IPC_HANDLE_SIZE.  Does moving that info ofi.h or another common file remove the current dependency from ofi_mr.h on ofi_hmem.h?  I think it does.

Reply:
With some extra modifications, yes, I can make that work. I had to move this to make it work:
```
#define MAX_IPC_HANDLE_SIZE	64

/*
 * This structure is part of the
 * the shm communication protocol
 * defined in ofi_shm.h.
 * Please make sure the SMR_VERSION are
 * bumped and SMR_CMD_SIZE are large
 * enough, for any changes in this
 * structure.
 */
struct ipc_info {
	uint64_t	iface;
	uint64_t	base_addr;
	uint64_t	base_length;
	uint64_t	device;
	uint64_t	offset;
	uint8_t		ipc_handle[MAX_IPC_HANDLE_SIZE];
};
```

Do you think ofi.h is the best place for it or would you like it somewhere else?

Reply:
Why did you need to move ipc_info?  I don't know that ofi.h is the best place, but looks like it without introducing a new header file.

Reply:
I agree moving `#define MAX_IPC_HANDLE_SIZE	64` to ofi.h and adding `#include <ofi.h>` to ofi_hmem.h is the simplest solution.